core/worker.go

```diff
@@ -1678,7 +1679,23 @@ func (w *worker) fillTransactions(env *environment, primeTerminus *types.WorkObj
 		txByPriceAndNonce.SetHead(orderedTxs)
 
 		env.wo.Header().SetBaseFee(baseFee)
-		return w.commitTransactions(env, primeTerminus, block, &txByPriceAndNonce, false)
+		err := w.commitTransactions(env, primeTerminus, block, &txByPriceAndNonce, false)
+		if err != nil {
+			return err
+		}
+		// If the length of txs is not the same as the length of orderedTxs, escape
+		nonEtxsCount := 0
+		for _, tx := range env.txs {
+			if tx.Type() != types.ExternalTxType {
+				nonEtxsCount++
+			}
+		}
+
+		if len(orderedTxs) != nonEtxsCount {
+			return errors.New("length of ordered txs is not the same as the non etxs count")
```

Comment:
It's permissible that a Qi transaction might be thrown out of the ordered list if the UTXO was already spent right?

Reply:
@jdowning100 This is a fresh run of the worker, so it will run on the current set

Reply:
it holds the headermu lock, it calls the GeneratePendingHeader

Reply:
```
	w.hc.headermu.Lock()
	head, err := w.GeneratePendingHeader(block, true, finalTxSet, true)
	if err != nil {
		log.Global.Error("Error generating pending header", err)
		w.hc.headermu.Unlock()
		return
	}
	w.hc.headermu.Unlock()
	``` in worker.go

Reply:
I mean before that, when you get the txs from the txpool and order them, some of them could be invalid by the time you call GeneratePendingHeader

Reply:
yeah, then we reject the whole run right? Is that a problem ?